File: csharp/src/CodexWebMcp/Cache.cs
using System;
using System.IO;
using System.Security.Cryptography;
using System.Text;
using System.Text.Json;
using System.Text.Json.Serialization;
using System.Threading;
using System.Threading.Tasks;

namespace CodexWebMcp;

internal sealed record CacheEntry(string Content, string Timestamp);

[JsonSourceGenerationOptions(WriteIndented = false)]
[JsonSerializable(typeof(CacheEntry))]
internal partial class CacheJsonContext : JsonSerializerContext { }

public static class Cache
{
    /// <summary>
    /// SHA-256 hex of "{tool}::{payload}", truncated to 32 chars.
    /// </summary>
    public static string KeyHash(string tool, string payload)
    {
        var combined = $"{tool}::{payload}";
        Span<byte> hash = stackalloc byte[32];
        var bytes = Encoding.UTF8.GetBytes(combined);
        SHA256.HashData(bytes, hash);

        // Hex encode (lower-case), then truncate to 32 chars.
        var sb = new StringBuilder(64);
        foreach (var b in hash)
            sb.Append(b.ToString("x2"));
        return sb.ToString(0, 32);
    }

    private static string FilePath(string tool, string payload)
    {
        var hash = KeyHash(tool, payload);
        return Path.Combine(Config.CacheSubdir, $"{tool}-{hash}.json");
    }

    public static async Task<string?> GetAsync(string tool, string payload, int ttlSeconds, CancellationToken ct = default)
    {
        try
        {
            var path = FilePath(tool, payload);
            if (!File.Exists(path)) return null;

            var mtime = File.GetLastWriteTimeUtc(path);
            var age = DateTime.UtcNow - mtime;
            if (age.TotalSeconds > ttlSeconds)
                return null;

            await using var fs = File.OpenRead(path);
            var entry = await JsonSerializer.DeserializeAsync(
                fs, CacheJsonContext.Default.CacheEntry, ct);
            return entry?.Content;
        }
        catch (Exception ex)
        {
            Console.Error.WriteLine($"cache.get error: {ex.Message}");
            return null;
        }
    }

    public static async Task PutAsync(string tool, string payload, string content, CancellationToken ct = default)
    {
        try
        {
            // Ensure directory exists (Config.CacheSubdir creates it, but be defensive)
            Directory.CreateDirectory(Config.CacheSubdir);

            var path = FilePath(tool, payload);
            var entry = new CacheEntry(content, DateTime.UtcNow.ToString("o"));

            // Atomic-ish write: write to temp then move
            var tmp = path + ".tmp";
            await using (var fs = File.Create(tmp))
            {
                await JsonSerializer.SerializeAsync(fs, entry, CacheJsonContext.Default.CacheEntry, ct);
            }

            if (File.Exists(path))
            {
                try { File.Delete(path); } catch { /* ignore */ }
            }
            File.Move(tmp, path);
        }
        catch (Exception ex)
        {
            Console.Error.WriteLine($"cache.put error: {ex.Message}");
        }
    }
}

File: java/src/main/java/codexwebmcp/Cache.java
package codexwebmcp;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.node.ObjectNode;

import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.time.Instant;

public final class Cache {
    private static final ObjectMapper MAPPER = new ObjectMapper();

    private Cache() {}

    public static String keyHash(String tool, String payload) {
        try {
            MessageDigest md = MessageDigest.getInstance("SHA-256");
            byte[] hash = md.digest((tool + "::" + payload).getBytes(StandardCharsets.UTF_8));
            StringBuilder sb = new StringBuilder(64);
            for (byte b : hash) sb.append(String.format("%02x", b));
            return sb.substring(0, 32);
        } catch (NoSuchAlgorithmException e) {
            throw new IllegalStateException("SHA-256 unavailable", e);
        }
    }

    private static Path filePath(String tool, String payload) throws IOException {
        return Config.cacheSubdir().resolve(tool + "-" + keyHash(tool, payload) + ".json");
    }

    public static String get(String tool, String payload, int ttlSeconds) {
        try {
            Path p = filePath(tool, payload);
            if (!Files.exists(p)) return null;
            long mtime = Files.getLastModifiedTime(p).toMillis();
            long ageMs = System.currentTimeMillis() - mtime;
            if (ageMs > ttlSeconds * 1000L) return null;
            byte[] data = Files.readAllBytes(p);
            ObjectNode node = (ObjectNode) MAPPER.readTree(data);
            if (node == null) return null;
            return node.has("content") ? node.get("content").asText() : null;
        } catch (IOException | ClassCastException e) {
            System.err.println("[codex-web-mcp] cache.get error: " + e.getMessage());
            return null;
        }
    }

    public static void put(String tool, String payload, String content) {
        try {
            Path p = filePath(tool, payload);
            ObjectNode entry = MAPPER.createObjectNode();
            entry.put("content", content);
            entry.put("timestamp", Instant.now().toString());
            byte[] bytes = MAPPER.writeValueAsBytes(entry);
            Path tmp = p.resolveSibling(p.getFileName().toString() + ".tmp");
            Files.write(tmp, bytes);
            try {
                Files.move(tmp, p, java.nio.file.StandardCopyOption.REPLACE_EXISTING,
                    java.nio.file.StandardCopyOption.ATOMIC_MOVE);
            } catch (java.nio.file.AtomicMoveNotSupportedException ex) {
                Files.move(tmp, p, java.nio.file.StandardCopyOption.REPLACE_EXISTING);
            }
        } catch (IOException e) {
            System.err.println("[codex-web-mcp] cache.put error: " + e.getMessage());
        }
    }
}
